nexus/db-queries/src/db/datastore/region.rs

```diff
@@ -240,6 +241,76 @@ impl DataStore {
         Ok(dataset_and_regions)
     }
 
+    /// Allocate new regions for a snapshot volume
+    #[allow(clippy::too_many_arguments)]
+    pub async fn arbitrary_region_allocate_for_snapshot(
```

Comment:
Nit: What's the word `arbitrary_` doing for us in these functions?

Reply:
The original function (before adding the redundancy level as an parameter) was `disk_region_allocate`, which allocated `REGION_REDUNDANCY_THRESHOLD` regions. `arbitrary_region_allocate` was meant to accept an arbitrary redundancy parameter. The `_direct` suffix was used to differentiate the function that accepted

```rust
        disk_source: &params::DiskSource,
        size: external::ByteCount,
```

and

```rust
        block_size: u64,
        blocks_per_extent: u64,
        extent_count: u64,
```

Adding `_snapshot_` to the name in this PR is further meant to differentiate when allocating for snapshot volumes...

Maybe something like builder pattern could work here instead:

```rust
        let query = crate::db::queries::region_allocation::AllocationQuery()
            .for_disk(volume_id)
            .with_disk_source(disk_source)
            .with_size(size)
            .with_allocation_strategy(allocation_strategy)
            .with_redundancy(3)
            .build()?;

        let query = crate::db::queries::region_allocation::AllocationQuery()
            .for_snapshot(volume_id, snapshot_id)
            .with_block_size(block_size)
            .with_blocks_per_extent(blocks_per_extent)
            .with_extent_count(extent_count)
            .with_redundancy(1)
            .build()?;
```

Reply:
I do like the idea of a builder pattern, if it doesn't add too much overhead. I definitely did not grok that `arbitrary` referred to the "redundancy" aspect from the function invocation.

Reply:
Follow-up from rust huddle: seems like builder pattern might be overkill (dynamically verifying the right params kinda sucks, because we'd introduce a new error that should never happen, and statically verifying params would result in type explosion). However, using a combinations of structs / enums to group together argument types does seem sufficient to reduce the calling complexity here too.

Reply:
This ended up working very well, see 9c4487b0d

Reply:
Ship it!